src/ppopt/utils/region_overlap_utils.py

```diff
@@ -0,0 +1,184 @@
+from itertools import combinations
+
+import numpy
+from ..mpmilp_program import MPMILP_Program
+from ..critical_region import CriticalRegion
+
+from .mpqp_utils import get_bounds_1d, is_full_dimensional_1d
+
+
+def reduce_overlapping_critical_regions_1d(program: MPMILP_Program, regions: list) -> list:
+    """
+    This adjusts critical regions for the 1d-parameter case, so that there are no overlaps.
+
+    :param program: the MPMILP_Program to be solved
+    :param regions: a list of CriticalRegion objects
+    :return: a list of non-overlapping CriticalRegion objects
+    """
+
+    # guard around the statement
+    if program.num_t() != 1:
+        raise ValueError('reduce_overlapping_critical_regions_1d  requires a 1d-parameter problem')
+
+    overlaps_remaining, region_added, regions = identify_overlaps(program, regions)
+    while region_added:
+        possible_dual_degeneracy, region_added, regions = identify_overlaps(program, regions)
+        overlaps_remaining = overlaps_remaining or possible_dual_degeneracy
+
+    return regions, overlaps_remaining
+
+
+def identify_overlaps(program, regions):
+    # In a first step, we identify regions that are fully or partially overlapping If a region is fully contained in
```

Comment:
I agree with Coday's analysis here, this function is to complex in the sense there is a lot of repeated logic here and that you can refactor some of the logic to get higher reusability.